### PR TITLE
BRS for APIM to enable secure access 

### PR DESCRIPTION
**Pull request by pranchalsomani**

### CODE REVIEW

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+
+        $apiManagementWithoutHttpsEnabled | ForEach-Object {
+            $apiManagementservice = $_
+            $resourceGroupName = $_.ResourceGroupName
+           
+            Write-Host "Enabling HTTPS url scheme for API(s) of API Management Service: Resource ID - $($_.ResourceId), Resource Group Name - $($_.ResourceGroupName), Resource Name - $($_.ResourceName)" -ForegroundColor $([Constants]::MessageType.Warning)
+            
+            try
+            {   #To hold name of API(s) which are remediated and skipped.        
+                $listOfApiRemediated = @()
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:393]: Please change this and the below line to listOfApi**s***.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+                $listOfApiSkipped = @()
+                
+                $apiMgmtContext = New-AzAPIManagementContext -ResourceGroupName $_.ResourceGroupName -ServiceName $_.ResourceName
+                $apiMgmt = Get-AzAPIManagementAPI -Context $apiMgmtContext
+                $apiManagementservice.ListOfAPIWithDisabledHttps = $apiManagementservice.ListOfAPIWithDisabledHttps -split ","
+                $apiMgmt | ForEach-Object { 
+                    if($apiManagementservice.ListOfAPIWithDisabledHttps -contains  $_.Name)
+                    {
+                        Set-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId -Protocols @("Https") -ErrorAction SilentlyContinue
+                        $output = (Get-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId  -ErrorAction SilentlyContinue).Protocols
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:403]: Do we really need to execute another command to get the status? Please check if `Set-AzAPIManagementAPI` returns the state of the resource post the update.
Else, we can rely on the output of this command to determine if the command was a success/failure - Unless, you saw an issue where `Set-AzAPIManagementAPI()` succeeded, but, `Get..` returned a different output than the `Set..`.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+                $apiManagementservice.ListOfAPIWithDisabledHttps = $apiManagementservice.ListOfAPIWithDisabledHttps -split ","
+                $apiMgmt | ForEach-Object { 
+                    if($apiManagementservice.ListOfAPIWithDisabledHttps -contains  $_.Name)
+                    {
+                        Set-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId -Protocols @("Https") -ErrorAction SilentlyContinue
+                        $output = (Get-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId  -ErrorAction SilentlyContinue).Protocols 
+                   
+                        if($output -eq "Https")
+                        {
+
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:407]: Please delete the stray newline.
`$output` will be a list of protocols. Please confirm this.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+                            $listOfApiRemediated += $_.Name
+                        }
+                        else
+                        {
+                            $listOfApiSkipped += $_.Name
+                        }
+                    }
+                }  
+
+                if($listOfApiRemediated.count -ne '0')
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:417]: `List.Count` should be returning an integer. Please check. This needs to be done in the below `if` block as
 well.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+                {
+                    $listOfApiRemediated = $listOfApiRemediated -join ","
+                    $apiManagementServicesRemediated += $apiManagementservice |Select-Object @{N='ResourceID';E={$_.ResourceId}},
+                                                                                            @{N='ResourceGroupName';E={$_.ResourceGroupName}},
+                                                                                            @{N='ResourceName';E={$_.ResourceName}},
+                                                                                            @{N='ListOfAPIRemediated';E={$listOfApiRemediated}}
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:423]: `ListOfAPIRemediated` to be replaced with `ListOfAPIsRemediated`. This needs to be done below as well - in the below `if` and `catch` blocks.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+                        else
+                        {
+                            $listOfApiSkipped += $_.Name
+                        }
+                    }
+                }  
+
+                if($listOfApiRemediated.count -ne '0')
+                {
+                    $listOfApiRemediated = $listOfApiRemediated -join ","
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:419]: Also, even if there is one `skipped` API in an APIM resource, the resource remediation should be treated as an error/incomplete.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 7]
```diff
@@ -0,0 +1,769 @@
...
+    }
+    else 
+    {
+        Write-Host $([Constants]::DoubleDashLine)
+        Write-Host "[Step 4 of 4] API Management services details have been backed up to $($backupFile). Please review before remediating them." -ForegroundColor $([Constants]::MessageType.Info)
+        Write-Host "Run the same command with -FilePath $($backupFile) and without -DryRun, to enable HTTPS url scheme for API(s) of all API Management services listed in the file." -ForegroundColor $([Constants]::MessageType.Info)
+    }
+}
+
+function Disable-HTTPSUrlSchemeForAPIsOfAPIManagementServices
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:495]: Please fix this after finalizing the file name / remediation command.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 8]
```diff
@@ -0,0 +1,769 @@
...
+        .INPUTS
+        None. You cannot pipe objects to Disable-HTTPSUrlSchemeForAPIsOfAPIManagementServices.
+
+        .OUTPUTS
+        None. Disable-HTTPSUrlSchemeForAPIsOfAPIManagementServices does not return anything that can be piped and used as an input to another command.
+
+        .EXAMPLE
+        PS> Disable-HTTPSUrlSchemeForAPIsOfAPIManagementServices -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -FilePath C:\AzTS\Subscriptions\00000000-xxxx-0000-xxxx-000000000000\202109131040\EnableHTTPSUrlSchemeForAPIsOfAPIManagementServices\RemediatedAPIManagementServices.csv
+
+        .EXAMPLE
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:526]: How is this command different from the above?

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 9]
```diff
@@ -0,0 +1,769 @@
...
+    if (-not (Test-Path -Path $FilePath))
+    {
+        Write-Host "ERROR: Input file - $($FilePath) not found. Exiting..." -ForegroundColor $([Constants]::MessageType.Error)
+        break
+    }
+
+    Write-Host "Fetching all API Management services from $($FilePath)" -ForegroundColor $([Constants]::MessageType.Info)
+    $validAPIManagementServiceDetails = @()
+    $apiManagementServiceDetails = Import-Csv -LiteralPath $FilePath
+    $apiManagementServiceDetails | ForEach-Object{
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:610]: Can this filtering be done similar to L251. That is, combine L608 and L610.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 10]
```diff
@@ -0,0 +1,769 @@
...
+    if ($totalAPIManagementResources -eq 0)
+    {
+        Write-Host "No API Management services found. Exiting..." -ForegroundColor $([Constants]::MessageType.Update)
+        break
+    }
+
+    Write-Host "Found $($totalAPIManagementResources) API Management Service(s)." -ForegroundColor $([Constants]::MessageType.Update)
+    
+    # Back up snapshots to `%LocalApplicationData%'.
+    $backupFolderPath = "$([Environment]::GetFolderPath('LocalApplicationData'))\AzTS\Remediation\Subscriptions\$($context.Subscription.SubscriptionId.replace('-','_'))\$($(Get-Date).ToString('yyyyMMddhhmm'))\DisableHTTPSUrlSchemeForAPIsOfAPIManagementServices"
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:628]: Please fix the file name `Url` vs `HTTPS`, and also to reflect the other file name related modifications being done in the remediation command.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 11]
```diff
@@ -0,0 +1,769 @@
...
+
+    # Includes API Management services, to which, previously made changes were successfully rolled back.
+    $apiManagementServicesRolledBack = @()
+
+    # Includes API Management services that were skipped during roll back. There were errors rolling back the changes made previously.
+    $apiManagementServicesSkipped = @()
+
+
+    $validAPIManagementServiceDetails | ForEach-Object {
+        $apiManagementservice = $_
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:665]: Please rename the variable from `apiManagementservice` to `apiManagementService`.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 12]
```diff
@@ -0,0 +1,769 @@
...
+           
+        Write-Host "Enabling HTTPS url scheme for API(s) of API Management Service: Resource ID - $($_.ResourceId), Resource Group Name - $($_.ResourceGroupName), Resource Name - $($_.ResourceName)" -ForegroundColor $([Constants]::MessageType.Warning)
+            
+        try
+        {           
+            $listOfApiRolledBack = @()
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:671]: Please change this and the below variable to listOfApi**s***.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 13]
```diff
@@ -0,0 +1,769 @@
...
+            $listOfApiSkipped = @()
+            #$$apiManagementservice.ListOfAPIRemediated
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:673]: Please delete this if not required.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 14]
```diff
@@ -0,0 +1,769 @@
...
+            $apiMgmtContext = New-AzAPIManagementContext -ResourceGroupName $_.ResourceGroupName -ServiceName $_.ResourceName
+            $apiMgmt = Get-AzAPIManagementAPI -Context $apiMgmtContext
+            $apiManagementservice.ListOfAPIRemediated  = $apiManagementservice.ListOfAPIRemediated -split ","
+            $apiMgmt | ForEach-Object {      
+                if($apiManagementservice.ListOfAPIRemediated -contains  $_.Name)
+                {
+                    Set-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId -Protocols @("Http") -ErrorAction SilentlyContinue
+                    $output = (Get-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId  -ErrorAction SilentlyContinue).Protocols 
+                        
+                    if($output -eq "Http")
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:683]: `$output` will be a list of protocols. Please confirm this.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 15]
```diff
@@ -0,0 +1,769 @@
...
+            $listOfApiSkipped = @()
+            #$$apiManagementservice.ListOfAPIRemediated
+            $apiMgmtContext = New-AzAPIManagementContext -ResourceGroupName $_.ResourceGroupName -ServiceName $_.ResourceName
+            $apiMgmt = Get-AzAPIManagementAPI -Context $apiMgmtContext
+            $apiManagementservice.ListOfAPIRemediated  = $apiManagementservice.ListOfAPIRemediated -split ","
+            $apiMgmt | ForEach-Object {      
+                if($apiManagementservice.ListOfAPIRemediated -contains  $_.Name)
+                {
+                    Set-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId -Protocols @("Http") -ErrorAction SilentlyContinue
+                    $output = (Get-AzAPIManagementAPI -Context $apiMgmtContext -ApiId $_.ApiId  -ErrorAction SilentlyContinue).Protocols
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:681]: Do we really need to execute another command to get the status? Please check if `Set-AzAPIManagementAPI` returns the state of the resource post the update.
Else, we can rely on the output of this command to determine if the command was a success/failure - Unless, you saw an issue where `Set-AzAPIManagementAPI()` succeeded, but, `Get..` returned a different output than the `Set..`.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 16]
```diff
@@ -0,0 +1,769 @@
...
+                        $listOfApiRolledBack += $_.Name
+                    }
+                    else
+                    {
+                        $listOfApiSkipped += $_.Name
+                    }
+                }
+            } 
+
+            if($listOfApiRolledBack.count -ne '0')
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:695]: `List.Count` should be returning an integer. Please check. This needs to be done in the below `if` block as
 well.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 17]
```diff
@@ -0,0 +1,769 @@
...
+            {
+                $listOfApiRolledBack = $listOfApiRolledBack -join ","
+                $apiManagementServicesRolledBack += $apiManagementservice |Select-Object @{N='ResourceID';E={$_.ResourceId}},
+                                                                                        @{N='ResourceGroupName';E={$_.ResourceGroupName}},
+                                                                                        @{N='ResourceName';E={$_.ResourceName}},
+                                                                                        @{N='ListOfAPIRolledBack';E={$listOfApiRolledBack}}
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:701]: `ListOfAPIRolledBack` to be replaced with `ListOfAPIsRolledBack`. A similar change needs to be done for the `skipped` case below.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 18]
```diff
@@ -0,0 +1,769 @@
...
+            }
+
+            if($listOfApiSkipped.count -ne '0')
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:704]: Even if there is one `skipped` API in an APIM resource, the resource rollback should be treated as an error/incomplete.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 19]
```diff
@@ -0,0 +1,769 @@
...
+        }
+        catch
+        {
+            $listOfApiSkipped = $listOfApiSkipped -join ","
+            $apiManagementServicesSkipped += $apiManagementservice | Select-Object @{N='ResourceID';E={$_.ResourceId}},
+                                                                                    @{N='ResourceGroupName';E={$_.ResourceGroupName}},
+                                                                                    @{N='ResourceName';E={$_ResourceName}},
+                                                                                    @{N='ListOfAPISkipped';E={$listOfApiSkipped}}
+            Write-Host "Error enabling HTTPS url scheme on the API Management service. Error: $($_)" -ForegroundColor $([Constants]::MessageType.Error)
+            Write-Host "Skipping this API Management Service. HTTPS url scheme will not be enabled for any of the API(s) of this service." -ForegroundColor $([Constants]::MessageType.Error)
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:721]: We are disabling HTTPS here.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 20]
```diff
@@ -0,0 +1,769 @@
...
+        Write-Host $([Constants]::DoubleDashLine)
+        Write-Host "Rollback Summary:`n" -ForegroundColor $([Constants]::MessageType.Info)
+        
+        if ($($apiManagementServicesRolledBack | Measure-Object).Count -gt 0)
+        {
+            Write-Host "HTTPS url scheme successfully disabled for the API(s) of following API Management Service(s):" -ForegroundColor $([Constants]::MessageType.Update)
+            $apiManagementServicesRolledBack | Format-Table -Property ResourceGroupName , ResourceName , ListOfAPIRolledBack
+
+            # Write this to a file.
+            $apiManagementServiceRolledBackFile = "$($backupFolderPath)\RolledBackApiManagementServices.csv"
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:737]: Ideally it is this file name convention that needs to be followed everywhere. Please use this as a reference for your changes.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

File: Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1 [hunk 21]
```diff
@@ -0,0 +1,769 @@
...
+    To roll back:
+        1. Download the script.
+        2. Load the script in a PowerShell session. Refer https://aka.ms/AzTS-docs/RemediationscriptExcSteps to know more about loading the script.
+        3. Execute the script to disable HTTPS url scheme on the API(s) of API Management services in the Subscription. Refer `Examples`, below.
+
+# Examples:
+    To remediate:
+        1. To review the API Management services in a Subscription that will be remediated:
+    
+           Enable-HTTPSUrlSchemeForAPIsOfAPIManagementServices -SubscriptionId 00000000-xxxx-0000-xxxx-000000000000 -PerformPreReqCheck -DryRun
```
[the remark this answers is not in the document]
Comment (pranchalsomani) [reply Scripts/RemediationScripts/Remediate-EnableHTTPSUrlSchemeForAPIManagementServices.ps1:41]: done

File: Scripts/RemediationScripts/Remediate-EnableHTTPSForAPIsInAPIManagementServices.ps1 [hunk 22]
```diff
@@ -0,0 +1,760 @@
...
+
+        $apiManagementServicesWithoutHttpsEnabled| ForEach-Object {
+            $apiManagementService = $_
+            $resourceGroupName = $_.ResourceGroupName
+           
+            Write-Host "Enabling HTTPS URL Scheme for API(s) of API Management Service: Resource ID - $($_.ResourceId), Resource Group Name - $($_.ResourceGroupName), Resource Name - $($_.ResourceName)" -ForegroundColor $([Constants]::MessageType.Warning)
+            
+            try
+            {   #To hold name of API(s) which are remediated and skipped.        
+               $listOfApisRemediated = @()
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSForAPIsInAPIManagementServices.ps1:389]: NIT: Please fix the indentation.

File: Scripts/RemediationScripts/Remediate-EnableHTTPSForAPIsInAPIManagementServices.ps1 [hunk 23]
```diff
@@ -0,0 +1,760 @@
...
+                $listOfApisSkipped = @()
+                
+                $apiMgmtContext = New-AzAPIManagementContext -ResourceGroupName $ResourceGroupName -ServiceName $_.ResourceName
```
Comment (gvaradarajan-msft) [inline Scripts/RemediationScripts/Remediate-EnableHTTPSForAPIsInAPIManagementServices.ps1:392]: The variable in L383 is `$resourceGroupName` - starts with a lower case, unlike `$ResourceGroupName`.

**Review by gvaradarajan-msft — CHANGES_REQUESTED** [review]

**Review by gvaradarajan-msft — APPROVED** [review]

**Pull request merged**